Tutorial: Write Your First Flink SQL program (#18055)
apache/flink: docs/content/docs/try-flink/write_flink_program_with_sql.md

- use the Flink SQL client to submit queries 
- consume a data source with Flink SQL
- run a continuous query on a stream of data
- use Flink SQL to write out results to persistent storage 

## Prerequisites 

You only need to have basic knowledge of SQL to follow along.

## Step 1: Start the Flink SQL client 

The [SQL Client]({{< ref "docs/dev/table/sqlClient" >}}) is bundled in the regular Flink distribution and can be run out-of-the-box. It requires only a running Flink cluster where table programs can be executed (since Flink SQL is a thin abstraction over the Table API). 

There are many ways to set up Flink but you will run it locally for the purpose of this tutorial. [Download Flink]({{< ref "docs/try-flink/local_installation#downloading-flink" >}}) and [start a local cluster]({{< ref "docs/try-flink/local_installation#starting-and-stopping-a-local-cluster" >}}) with one worker (the TaskManager).  

The scripts for the SQL client are located in the `/bin` directory of Flink. You can start the client by executing:

```sh
./bin/sql-client.sh
```

You should see something like this:

{{< img src="/fig/try-flink/flink-sql.png" alt="Flink SQL client" >}}

> ## Step 2: Set up a data source with flink-faker

Reviewer: for a beginner tutorial, the user needs to download many things. wouldn't it be better to use Flink's datagen which is built-in and doesn't require additional steps.

Author: I wanted to find a balance between realistic and simplicity.  Do you think that the datagen connector is as flexible as flink-faker and can create realistic examples?

Reviewer 2: If you really wanna do a realistic example, then I suggest to go with a full fledged connector like kafka or filesystem, instead of flink-faker, so you can also show the concept of connector and format.

Reviewer 3: I fear if we use datagen the examples will be significantly less compelling.

Reviewer 3: @slinkydeveloper I get where you're coming from, but using real connectors is problematic.

With Kafka we'd either have to force folks to install kafka (not gonna do that), use docker (also not very workable, unfortunately), or we somehow set up an embedded kafka (again, no). Plus we'd need to somehow generate data into kafka. Besides, we already have examples based on docker-ized kafka, and they haven't worked out very well.

Using the filesystem might be a great idea, but I believe we want this to be a streaming example, and afaik we don't yet have a streaming filesystem table source.

Reviewer 2: > Using the filesystem might be a great idea, but I believe we want this to be a streaming example, and afaik we don't yet have a streaming filesystem table source. Glad you're working on https://issues.apache.org/jira/browse/FLINK-20286.

Exactly, and the streaming filesystem table source should be ready for the next release.

Another thing I personally don't like is that flink-faker is not an official Apache project, but instead it's a community project, hence I don't feel comfortable using it here in the official docs.

I personally prefer that either you use datagen or connector files, as IMHO flink-faker is not a good connector example.

Reviewer 3: I don't see how to create a good streaming SQL demo/tutorial experience on 1.14 without using flink-faker.

Let's assume we'll have a filesystem table source in 1.15 -- how will we use it in tutorials like this one? I'm guessing we can somehow generate a file containing a non-trivial amount of interesting data, and then host that file somewhere. Would we put that at some convenient URI -- maybe in an S3 bucket, or something similar -- and configure a table that points directly there?

Author: Personally, I am in favor of keeping flink-faker for now to get the piece published and then iterating on it.  Even though it is not an official Apache project, it was created by someone from Ververica and I feel like it doesn't need that much maintenance based on what it does.  Although I could be wrong. :p

Reviewer 3: An alternative would be to switch over to using a file now, and rework the tutorial to be batch only until 1.15, after which it could be expanded to show both batch and streaming.

But I'm still unsure about the logistics of using a file. Ideally such a file should be hosted somewhere under the control of the project -- so maybe in some github repo.

Reviewer 2: I think even a simple file committed here is enough, something like a single csv file or 1000 lines

Reviewer 4: We use the flink-faker extensively in the Ververica SQL recipes. I would +1 for using the flink-faker here too. You could even make an argument that for proper tutorials like this, we should have a flink-faker in the Flink repo itself (but that's another discussion)

Like with any Flink program, you will need a data source to connect to so that Flink can process it. There are many popular data sources but for the interest of this tutorial, you will be using [flink-faker](https://github.com/knaufk/flink-faker).  This custom [table source]({{< ref "docs/connectors/table/overview" >}}) is based on [Java Faker](https://github.com/DiUS/java-faker) and can generate fake data continuously in memory and in a realistic format. 

> Java Faker is a tool for generating this data and flink-faker exposes that as a source in Flink by implementing the [DynamicTableSource interface]({{< javadoc file="org/apache/flink/table/connector/source/DynamicTableSource.html" name="DynamicTableSource">}}). The dynamic table source has the logic of how to create a table source (in this case, from flink-faker), and then by adding a [factory]({{< javadoc file="org/apache/flink/table/factories/DynamicTableSourceFactory.html" name="DynamicTableSourceFactory">}}) for it you can expose it in the SQL API by referencing it with `"connector" = "faker"`.

Reviewer: I wouldn't reference JavaDocs in beginner tutorials. It is actually not that important to understand `DynamicTableSource` or `DynamicTableSourceFactory` as those are second level API. If at all you can reference the documentation of https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sourcessinks/.

Author: What would be considered first level APIs?

The next step is to make `flink-faker` discoverable by the Flink SQL client by following these [instructions](https://github.com/knaufk/flink-faker#adding-flink-faker-to-flink-sql-client).

Once you have done that, create a table using this table source to confirm that the factory is loaded by executing the following query in the SQL client: 

```sql
CREATE TABLE test WITH ('connector' = 'faker');
```

> If you see `[INFO] Execute statement succeed.`, then the table source has been loaded correctly.

Reviewer: This doesn't mean that the table source has been loaded correctly. It only means that a metadata object has been added to the catalog. whether the options are correct and a connector can be found will be checked when accessing the table.

In other words: `A table definition has been stored in the in-memory catalog successfully`

Btw a schema is missing in this statement. This looks more like a bug to me and should actually not be allowed.

Author: To work around the schema problem, could I do `CREATE TEMPORARY TABLE` instead?

You are ready to start writing your first Flink program with SQL. 

## Step 3: Consume the data via SQL

For this tutorial, you are going to create a table that models a [Twitch](https://www.twitch.tv) gaming stream. This table will contain the following fields: user_name, game_name,viewer_count, started_at, location, and a timestamp.

> Use the [DDL syntax](https://www.geeksforgeeks.org/sql-ddl-dql-dml-dcl-tcl-commands/) `CREATE TABLE` to create this table containing these fields. You will also use the `WITH` clause to configure the connector (i.e. flink-faker). 

Reviewer: we should link to other websites. the links get out of date quickly and require maintenance overhead for us. we should (and also have) our own DDL page for linking.

Execute the following query in the SQL client:

```sql
CREATE TEMPORARY TABLE twitch_stream (
  `user_name` VARCHAR(2147483647),
  `game_name` VARCHAR(2147483647),
  `viewer_count` INT,
  `started_at` TIMESTAMP_LTZ(3),

>   `location` VARCHAR(2147483647),

Reviewer: never use `VARCHAR(2147483647)` use `STRING`.

Author: Out of curiosity, why is that?

Reviewer 2: :+1: with Timo, we are discouraging `VARCHAR` as its a type we inherit from old databases. It doesn't make much sense in streaming, and especially in Flink SQL, we support it only for legacy.

Flink does not provide its own storage system but instead offers many sink connectors you can use to write table updates to external systems.

To write a table to a CSV file, you can use the [FileSystem connector]({{< ref "docs/connectors/table/filesystem" >}}) that is built into Flink. This connector supports row-encoded and bulk-encoded formats. Row-encoded formats such as CSV and JSON write one row at a time to a file. Bulk-encoded formats collect a batch of rows in memory and organize them in a storage-and-scan-efficient format before writing out the data. 

To store the results of the last example query (tumble window), you need to create a table that will store the results:

```sql
CREATE TABLE twitch_stream_viewer_count (
  `window_start` TIMESTAMP_LTZ(3),
  `window_end` TIMESTAMP_LTZ(3),
  `game_name` VARCHAR(2147483647),
  `TotalViewerCount` INT
) WITH (
  'connector' = 'filesystem',
  'path' = 'file:///path/to/directory/results.csv',
  'format' = 'csv'
)
```

Now insert the streaming SQL results into the file system table:

```sql
INSERT INTO twitch_stream_viewer_count 
SELECT 
  TBD
FROM twitch_stream;

> ```

Author: @alpinegizmo Can you help me with the insert query into the sink?

The scripts for the SQL client are located in the `/bin` directory of Flink. You can start the client by executing:

```sh
./bin/sql-client.sh
```

You should see something like this:

{{< img src="/fig/try-flink/flink-sql.png" alt="Flink SQL client" >}}

## Step 2: Set up a data source with flink-faker

Like with any Flink program, you will need a data source to connect to so that Flink can process it. There are many popular data sources but for the interest of this tutorial, you will be using [flink-faker](https://github.com/knaufk/flink-faker).  This custom [table source](https://nightlies.apache.org/flink/flink-docs-release-1.13/docs/connectors/table/overview/) is based on [Java Faker](https://github.com/DiUS/java-faker) and can generate fake data continuously in memory and in a realistic format. 

> Java Faker is a tool for generating this data and flink-faker exposes that as a source in Flink by implementing the [DynamicTableSource interface](https://nightlies.apache.org/flink/flink-docs-release-1.11/api/java/org/apache/flink/table/connector/source/DynamicTableSource.html). The dynamic table source has the logic of how to create a table source (in this case, from flink-faker), and then by adding a factory (https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/table/factories/DynamicTableSourceFactory.html) for it you can expose it in the SQL API by referencing it with `"connector" = "faker"`.

Author: ooops mistake

Dynamic tables are the fundamental concept behind Flink's SQL support for streaming data. While SQL makes it seem like you are querying a database (or static tables that represent batch data, the tables in Flink are actually dynamic tables that are defined by queries. So instead of running several different queries on the same set of data, you are continuously running one query on a dataset that keeps changing. 

Under the hood, the SQL client submits queries to Flink's JobManager, which works with the TaskManager to assign and monitor query tasks. Have a look at [Flink's architecture](https://nightlies.apache.org/flink/flink-docs-master/docs/concepts/flink-architecture/) for more detail. 

## Step 5: Filter the data

Now try to perform a filter operation on this data stream to specify a subset of the data by using the `WHERE` keyword.  

To find out what gaming streams started in the last 15 minutes, try this query:

```sql

> SELECT game_name, 
>        started_at,
>        proctime, 
>        TIMESTAMPDIFF(MINUTE,CAST(started_at AS TIMESTAMP(3)),CAST(proctime as TIMESTAMP(3)))
> FROM twitch_stream
> WHERE TIMESTAMPDIFF(MINUTE,CAST(started_at AS TIMESTAMP(3)),CAST(proctime as TIMESTAMP(3))) < 15;

Reviewer 3: ```suggestion
SELECT game_name, 
       started_at,
       proctime, 
       TIMESTAMPDIFF(MINUTE, started_at, proctime)
FROM twitch_stream
WHERE TIMESTAMPDIFF(MINUTE, started_at, proctime) < 15;
```

```

You should now see a new table with new datastream results.  

> Now try a [Top-N](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/queries/topn/) query to find out the top 10 most popular games.

Reviewer 3: ```suggestion
Now try a [Top-N](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/queries/topn/) query to find the 10 most popular games.
```

> Top-N queries identify the N smallest or largest values ordered by columns and is useful when you need to identify the top N items in a stream, or the bottom N items in a stream. 

Reviewer 3: ```suggestion
Top-N queries identify the N smallest or largest values (as ordered by some attribute of the table), and are useful when you need to identify the top (or bottom) N items in a stream. 
```

```sql
SELECT game_name, proctime, viewer_count
FROM (
    SELECT *, ROW_NUMBER() OVER (ORDER BY viewer_count DESC) AS ranking
    FROM (
        SELECT *
        FROM (
            SELECT 
                *,
                ROW_NUMBER() OVER (PARTITION BY game_name ORDER BY proctime DESC) AS ordering
            FROM twitch_stream
        ) WHERE ordering = 1
    )
) WHERE ranking <= 10;
```

Note that Flink uses the combination of an OVER window clause and a filter condition to express a Top-N query in order to filter through unbounded dataset. 

> You should now see a new table with these new datastream results.  

Reviewer 3: ```suggestion
You should now see a new dynamic table with the results from this query.  Notice how the top 10 games are constantly being revised as new data is processed.
```

For the last example, let's try a very commonly used function in streaming SQL: computing windowed aggregations.

Windows are the fundamental building blocks when it comes to processing infinite streams since they provide a way to split streams into finite chunks. 
Flink SQL offers several [window functions](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/queries/window-tvf/) that you can use.  

Let's explore how to apply one of these functions! To calculate an aggregation per game of the total viewer count for every 1 minute window, 
you can use the [tumble](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/queries/window-tvf/#tumble) function.

Tumbling windows can be thought of as mini-batches of aggregations over a non-overlapping window of time.

```sql
SELECT window_start,
       window_end,
       game_name,
       SUM(viewer_count) AS TotalViewerCount
  FROM TABLE(
    TUMBLE(TABLE twitch_stream, DESCRIPTOR(proctime), INTERVAL '1' MINUTE))
  GROUP BY window_start, window_end, game_name;
```

> The results will refresh every minute and calculates the sum of all viewer counts for each game from when you started running this query. 

Reviewer 3: ```suggestion
Fresh results will appear every minute, showing the sum of the viewer counts for each game.
```

## Step 7: Write the updated stream to persistant storage

Now that you have written a program with Flink SQL to process all that streaming data, you probably want to store it somewhere. Since you are dealing with unbounded data sets, you can think of it as maintaining materialized views (or snapshots of the data) in external storage systems.

> Flink does not provide its own storage system but offers many sink connectors to external systems for you to write table updates to.

Reviewer 3: ```suggestion
Flink does not provide its own storage system but instead offers many sink connectors you can use to write table updates to external systems.
```

> To write the table row results to a CSV file, you can use the [FileSystem connector](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/filesystem/) that is built into Flink. This connector supports row-encoded and bulk-encoded formats. Row-encoded formats such as CSV and JSON write each row individually to a file. Bulk-encoded formats collect a batch of rows in memory and organize them in a storage and scan efficient format before writing the data to a file. 

Reviewer 3: ```suggestion
To write a table to a CSV file, you can use the [FileSystem connector](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/filesystem/) that is built into Flink. This connector supports row-encoded and bulk-encoded formats. Row-encoded formats such as CSV and JSON write one row at a time to a file. Bulk-encoded formats collect a batch of rows in memory and organize them in a storage-and-scan-efficient format before writing out the data. 
```

To store the results of the last example query (tumble window), you need to create a table that will store the results:

```sql
CREATE TABLE twitch_stream_viewer_count (
  `window_start` TIMESTAMP_LTZ(3),
  `window_end` TIMESTAMP_LTZ(3),
  `game_name` VARCHAR(2147483647),
  `TotalViewerCount` INT
) WITH (
  'connector' = 'filesystem',
  'path' = 'file:///path/to/directory/results.csv',
  'format' = 'csv'
)
```

Now insert the streaming SQL results into the file system table:

```sql
INSERT INTO twitch_stream_viewer_count 
SELECT 
  TBD
FROM twitch_stream;
```

## Summary

> In this tutorial, you learned how to use Flink SQL to define a whole continuous data pipeline. While not designed initially with streaming semantics in mind, SQL is a timeless and powerful query language that you can write a full program with. Where Flink SQL differs from traditional database queries is that it works with dynamic tables and is continuously consuming rows as they arrive and producing updates to the tables.

Reviewer 3: ```suggestion
In this tutorial, you learned how to use Flink SQL to define a whole continuous data pipeline. While not designed initially with streaming semantics in mind, SQL is a timeless and powerful query language that you can write comp;lete programs. Where Flink SQL differs from traditional database queries is that it works with dynamic tables and is continuously consuming rows as they arrive and is continuously producing updates to the result tables.
```

All Flink SQL programs follow a similar pattern: define a table source, perform manipulations on the data, persist the data to storage.

> Moreover, you were introduced to some fundamental concepts when processing unbounded streams of data. You learned about [time attributes](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/concepts/time_attributes/) such as event time vs. processing time and how to deal with data order.  You also learned about [windows](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/operators/windows/) and how to apply that with Flink SQL on your data.  Windows are the key to turning neverending streams of data into manageable chunks for computation and are integral to any stream processor. 

Reviewer 3: I don't think we really learned anything about event time or out-of-order data.

```sql
SELECT * FROM twitch_stream;
```

You should see, on the console of the SQL client, a stream of data populating each of the defined fields of the twitch_stream table.  Notice that this table contains a stream of data and is what is known as a [dynamic table]({{< ref "docs/dev/table/concepts/dynamic_tables" >}}).  

Dynamic tables are the fundamental concept behind Flink's SQL support for streaming data. While SQL makes it seem like you are querying a database (or static tables that represent batch data, the tables in Flink are actually dynamic tables that are defined by queries. So instead of running several different queries on the same set of data, you are continuously running one query on a dataset that keeps changing. 

Under the hood, the SQL client submits queries to Flink's JobManager, which works with the TaskManager(s) to assign and monitor query tasks. Have a look at [Flink's architecture]({{< ref "docs/concepts/flink-architecture" >}}) for more detail. 

## Step 5: Filter the data

Now try to perform a filter operation on this data stream to specify a subset of the data by using the `WHERE` keyword.  

> To find out what gaming streams started in the last 15 minutes, try this query:
> 
> ```sql
> SELECT game_name, 
>        started_at,
>        proctime, 
>        TIMESTAMPDIFF(MINUTE, started_at, proctime)
> FROM twitch_stream
> WHERE TIMESTAMPDIFF(MINUTE, started_at, proctime) < 15;
> ```

Reviewer 3: Everything started in the last 15 minutes, so this example is both a bit complex looking, and the results aren't actually filtered at all. How about this instead?

```suggestion
To find gaming streams for games whose names end with an exclamation mark, try this query:

```sql
SELECT *
FROM twitch_stream
WHERE game_name LIKE '%!';
```
```

You should now see a new table with new datastream results.  

> 

Reviewer 3: top-n isn't really another example of filtering. Maybe insert a new Step 6 (and increment all of the later steps)?

Now try a [Top-N]({{< ref "docs/dev/table/sql/queries/topn" >}}) query to find the 10 most popular games.

Top-N queries identify the N smallest or largest values (as ordered by some attribute of the table), and are useful when you need to identify the top (or bottom) N items in a stream. 

```sql
SELECT game_name, proctime, viewer_count
FROM (
    SELECT *, ROW_NUMBER() OVER (ORDER BY viewer_count DESC) AS ranking
    FROM (
        SELECT *
        FROM (
            SELECT 
                *,
                ROW_NUMBER() OVER (PARTITION BY game_name ORDER BY proctime DESC) AS ordering
            FROM twitch_stream
        ) WHERE ordering = 1
    )
) WHERE ranking <= 10;
```

Note that Flink uses the combination of an OVER window clause and a filter condition to express a Top-N query in order to filter through unbounded dataset. 

You should now see a new dynamic table with the results from this query.  Notice how the top 10 games are constantly being revised as new data is processed.

## Step 6: Aggregate the data and learn about windowing 

> Now try an aggregate function (`COUNT`, `SUM`, `AVG`, `MIN`, `MAX`) to find the most popular games based on location with the following query: 

Reviewer 3: This query isn't doing anything to limit the results to only the most popular games.

WHERE location IS NOT NULL
GROUP BY location, game_name;
```

You should see a table result of continously changing data that tells you the most popular games for each country and the count for each game. 

For the last example, let's try a very commonly used function in streaming SQL: computing windowed aggregations.

Windows are the fundamental building blocks when it comes to processing infinite streams since they provide a way to split streams into finite chunks. 
Flink SQL offers several [window functions]({{< ref "docs/dev/table/sql/queries/window-tvf" >}}) that you can use.  

Let's explore how to apply one of these functions! To calculate an aggregation per game of the total viewer count for every 1 minute window, 
you can use the [tumble]({{< ref "docs/dev/table/sql/queries/window-tvf#tumble" >}}) function.

Tumbling windows can be thought of as mini-batches of aggregations over a non-overlapping window of time.

```sql
SELECT window_start,
       window_end,
       game_name,
       SUM(viewer_count) AS TotalViewerCount
  FROM TABLE(
    TUMBLE(TABLE twitch_stream, DESCRIPTOR(proctime), INTERVAL '1' MINUTE))
  GROUP BY window_start, window_end, game_name;
```

> Fresh results will appear every minute, showing the sum of the viewer counts for each game.

Reviewer 3: I found a minute to be rather a long time to wait. Maybe we should shorten the windows.

Flink does not provide its own storage system but instead offers many sink connectors you can use to write table updates to external systems.

To write a table to a CSV file, you can use the [FileSystem connector]({{< ref "docs/connectors/table/filesystem" >}}) that is built into Flink. This connector supports row-encoded and bulk-encoded formats. Row-encoded formats such as CSV and JSON write one row at a time to a file. Bulk-encoded formats collect a batch of rows in memory and organize them in a storage-and-scan-efficient format before writing out the data. 

To store the results of the last example query (tumble window), you need to create a table that will store the results:

```sql
CREATE TABLE twitch_stream_viewer_count (
  `window_start` TIMESTAMP_LTZ(3),
  `window_end` TIMESTAMP_LTZ(3),
  `game_name` VARCHAR(2147483647),
  `TotalViewerCount` INT
) WITH (
  'connector' = 'filesystem',
  'path' = 'file:///path/to/directory/results.csv',
  'format' = 'csv'
)
```

Now insert the streaming SQL results into the file system table:

```sql

> INSERT INTO twitch_stream_viewer_count 
> SELECT 
>   `window_start` TIMESTAMP_LTZ(3),
>   `window_end` TIMESTAMP_LTZ(3),
>   `game_name` VARCHAR(2147483647),
>   `TotalViewerCount` INT
> FROM twitch_stream;

Reviewer 3: This insert statement should be

```suggestion
INSERT INTO twitch_stream_viewer_count 
SELECT 
  `window_start` TIMESTAMP_LTZ(3),
  `window_end` TIMESTAMP_LTZ(3),
  `game_name` VARCHAR(2147483647),
  `TotalViewerCount` INT
FROM TABLE(
    TUMBLE(TABLE twitch_stream, DESCRIPTOR(proctime), INTERVAL '1' MINUTE))
GROUP BY window_start, window_end, game_name;
```

However, this INSERT INTO is problematic, because the filesystem sink actually creates a directory full of rolling files, and those files are only finalized during checkpointing, which isn't happening here. 

In my case, I ended up with a file named `/tmp/results.csv/.part-7662f404-d437-4351-8f64-7f71944fb9f4-0-0.inprogress.0354b5f8-0c13-4c40-9fcd-d5db0ca2bb44` containing the in-progress results. 

This part of the tutorial needs more thought -- I don't think we want to do anything this confusing.

Reviewer: @AHeise Is there anything in the near future that improve this? We need an option that reads and write single files for testing, demos etc. This is an early blocker that got reported multiple times.

Author: Perhaps leave it this way and then revisit it once we have that feature available?
